src/metatrain/utils/custom_loss.py

```diff
@@ -0,0 +1,430 @@
+from abc import ABC, ABCMeta, abstractmethod
+from typing import Any, Dict, Optional, Type
+
+import metatensor.torch as mts
+import torch
+from metatensor.torch import TensorMap
+from torch.nn.modules.loss import _Loss
+
+from metatrain.utils.data import TargetInfo
+
+
+class LossRegistry(ABCMeta):
+    """
+    Metaclass to auto-register :py:class:`LossInterface` subclasses.
+
+    Maintains a mapping from ``registry_name`` to the subclass type.
+    """
+
+    _registry: Dict[str, Type["LossInterface"]] = {}
+
+    def __new__(mcs, name, bases, attrs):
+        cls = super().__new__(mcs, name, bases, attrs)
+        # Skip the abstract base itself
+        if name != "LossInterface" and issubclass(cls, LossInterface):
+            # Use explicit registry_name if given, else snake_case from class name
+            key = getattr(cls, "registry_name", None)
+            if key is None:
+                key = "".join(
+                    f"_{c.lower()}" if c.isupper() else c for c in name
+                ).lstrip("_")
+            # only register the very first class under each key
+            mcs._registry.setdefault(key, cls)
```

Comment:
```suggestion
        if name != "LossInterface" and issubclass(cls, LossInterface):
          if "registry_name" not in attrs:
            raise TypeError(f"Class '{name}' must define a 'registry_name' class attribute")
```

Indentation might be off. It is good to have a fallback in general, but not if we only have a set number of loss functions in `metatrain`, since we can easily enforce a fixed name for each new loss (esp. with the `enum` suggestion).

If we end up with third party loss support, then this would make much more sense :)